Source/DafnyStandardLibraries/src/dafnystdlibs/Wrappers.md

```diff
@@ -0,0 +1,191 @@
+{
+  if k >= |s| then None else if s[k] == value then Some(k) else Find(s, k+1, value)
+}
+```
+
+It could be used in a method like this
+
+<!-- %check-resolve %use tmp-find.dfy -->
+```dafny
+method m(s: seq<int>) returns (r: Option<int>) {
+  var value: int;
+  // do various calculations
+  var index: int :- Find(s, 0, value);
+  // more calculations
+  return Some(value);
+}
+```
+
+You could just capture the result of `Find` using `:=` in a `Option<int>` variable and inspect it. But if the `None` condition is 
+generally a rare error, it is easy to forget to always check that each operation was successful. Instead, the `:-` changes the 
+control flow so that if a `None` value is returned from `Find`, the method immediately aborts, with the output value (which has
+the `Option<int>` type) getting that returned `None` value. So this operates something like exceptions.
+See the [reference manual](https://dafny.org/latest/DafnyRef/DafnyRef#sec-update-with-failure-statement)
+for more on the similarities and differences with exceptions.
+
+### Outcome
+
+`Outcome` is a variation on these types in which we mostly just care whether the operations succeeded or failed.
+It has no place to carry a value, just information about the error. 
+
+This datatype is most useful when used with methods with multiple return values.
+Then the first return value can be an `Outcome` to indicate success or failure, with the other outputs returning 
+appropriate values in the success situation.
+
+For example, if we have this method
+<!-- %check-resolve %save tmp-matches.dfy -->
+```dafny
+import opened DafnyStdLibs.Wrappers
+method FindAllMatches<T(==)>(s: seq<T>, value: T) returns (status: Outcome<string>, matches: seq<int>)
+```
+we can call it as
+<!-- %check-resolve %use tmp-matches.dfy -->
+```dafny
+method m(s: seq<int>, value: int) returns (r: Outcome<string>) {
+  var matches :- FindAllMatches(s, value);
+}
+```
+
+Notice that there is no left-hand-side for the first out-parameter. It does not carry a value: if the value is a `Fail`, the control
+flow will abruptly return; if is is a `Pass`, the first out-parameter is then discarded and the second is assigned to the remaining LHS.
+An `Outcome` serves as an automatically-checked error mechanism.
+
+A useful function in conjunction with Outcome values is `Need`. 
+This function takes a boolean condition and a value `e` of an error type `E` and returns an
+`Outcome<E>` value: `Pass` if the condition is true, `Fail(e)` if it is false. 
+It provides a simple runtime check that does not abort the program (like the `expect` statement does);
+rather it propagates the Dafny equivalent of an exception.
+
+<!-- %no-check -->
+```dafny
+:- Need(IsEverythingOK(), "failure in method M");
+```
+ 
+### Result
+
+A `Result` carries both a success value and failure information, with two separate, specifiable types. Its use is then very similar to `Option`.
+
+### Combining different FC-types in expressions
```

Comment:
While the acronym FC has been defined before (once, and was not used afterwards), I would not use it in the title here.

Reply:
Agreed about the titles, although "FC" is used a handful of times in the content.